src/MongoDB.Driver/Encryption/LibMongoCryptController.cs

```diff
@@ -0,0 +1,694 @@
+        {
+            var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+#if NETSTANDARD1_5
+            await socket.ConnectAsync(request.Endpoint, 443).ConfigureAwait(false);
+#else
+            await Task.Factory.FromAsync(socket.BeginConnect(request.Endpoint, 443, null, null), socket.EndConnect).ConfigureAwait(false);
+#endif
+
+            var validationCallback = new RemoteCertificateValidationCallback(AcceptAnyCertificate); // TODO: what validation needs to be done?
+
+            using (var networkStream = new NetworkStream(socket, ownsSocket: true))
+            using (var sslStream = new SslStream(networkStream, leaveInnerStreamOpen: false, validationCallback))
+            {
+                await sslStream.AuthenticateAsClientAsync(request.Endpoint).ConfigureAwait(false);
+
+                var requestBytes = request.Message.ToArray();
+                await sslStream.WriteAsync(requestBytes, 0, requestBytes.Length).ConfigureAwait(false);
+
+                var buffer = new byte[4096];
+                while (request.BytesNeeded > 0)
+                {
+                    var count = await sslStream.ReadAsync(buffer, 0, buffer.Length).ConfigureAwait(false);
+                    var responseBytes = new byte[count];
+                    Buffer.BlockCopy(buffer, 0, responseBytes, 0, count);
+                    request.Feed(responseBytes);
+                }
+            }
+        }
+
+        private void RestoreDbNodeInResponse(BsonDocument request, BsonDocument response)
+        {
+            if (request.TryGetElement("$db", out var db))
+            {
+                var result = response["result"].AsBsonDocument;
+                if (!result.Contains("$db"))
+                {
+                    result.Add(db);
+                }
+            }
+        }
+
+        private void ThrowIfNotAutoEncryption(CryptContext.StateCode state)
+        {
+            if (_encryptionMode != EncryptionMode.Auto)
+            {
+                throw new InvalidOperationException($"{state} is available only for auto encryption.");
+            }
+        }
+
+        private void ThrowIfNotInitialized()
+        {
+            if (!_initialized)
+            {
+                throw new InvalidOperationException($"{nameof(LibMongoCryptController)} must be initialized.");
+            }
+        }
+
+        private BsonValue UnwrapDecryptedValue(byte[] wrappedBytes)
+        {
+            var wrappedDocument = new RawBsonDocument(wrappedBytes);
+            return wrappedDocument["v"];
+        }
+
+        private BsonBinaryData UnwrapEncryptedValue(byte[] encryptedWrappedBytes)
+        {
+            var wrappedDocument = new RawBsonDocument(encryptedWrappedBytes);
+            var encryptedBytes = wrappedDocument["v"].AsByteArray;
+            return new BsonBinaryData(encryptedBytes, BsonBinarySubType.Encrypted);
```

Comment:
I don't see the need to create a new BsonBinaryData.

Lines 683-684 can just be:

    return wrappedDocument["v"].AsBsonBinaryData;